code/krepel/memory/allocation.d

```diff
@@ -0,0 +1,691 @@
+module krepel.memory.allocation;
+
+import krepel;
+import krepel.memory;
+import krepel.algorithm;
+import krepel.math : IsPowerOfTwo, IsEven, IsOdd;
+import Meta = krepel.meta;
+
+/// The global default allocator.
+__gshared ForwardAllocator!HeapMemory GlobalAllocator;
+
+/// Memory Features
+enum : ubyte
+{
+  SupportsAllocationOnly  = 0,
+  SupportsReallocation    = 1 << 0,
+  SupportsDeallocation    = 1 << 1,
+}
+
+/// Common functionality for all memory types.
+mixin template MemoryMixinTemplate(size_t InFeatures)
+{
+  alias ThisIsAMemoryType = typeof(this);
+  enum size_t Features = InFeatures;
+}
+
+/// Adds the Contains function to a memory type so it can be asked whether a
+/// given memory region or pointer belongs to this memory.
+mixin template MemoryContainsCheckMixin(alias Member)
+{
+  bool Contains(const MemoryRegion SomeRegion) const
+  {
+    return SomeRegion.ptr >= Member.ptr &&
+           SomeRegion.ptr - Member.ptr + SomeRegion.length < Member.length;
+  }
+
+  bool Contains(const ubyte* SomePointer) const
+  {
+    return SomePointer >= Member.ptr &&
+           SomePointer < Member.ptr + Member.length;
+  }
+}
+
+/// A template to determine whether the given type is a memory type.
+template IsSomeMemory(M)
+{
+  static if(is(M.ThisIsAMemoryType)) enum bool IsSomeMemory = true;
+  else                               enum bool IsSomeMemory = false;
+}
+
+/// An allocator that wraps a given memory type used to allocate single
+/// objects or entire arrays.
+struct ForwardAllocator(M)
+{
+  alias MemoryType = M;
+  static assert(IsSomeMemory!MemoryType);
+
+  MemoryType Memory;
+
+  /// Creates a new instance of T without constructing it.
+  T* NewUnconstructed(T)()
+  {
+    auto Raw = Memory.Allocate(T.sizeof, T.alignof);
+    if(Raw is null) return null;
+    assert(Raw.length >= T.sizeof);
+    auto Instance = cast(T*)Raw.ptr;
+    return Instance;
+  }
+
+  /// Free the memory occupied by Instance without destructing it.
+  /// Note: If the memory type used does not support deallocation
+  ///       (e.g. StackMemory), this function does nothing.
+  void DeleteUndestructed(T)(T* Instance)
```

Comment:
Well, it's not really raw for me... Raw implies you get a raw block of memory, but the returned block is already typed as `T`...

But why does _undestructed_ imply _destructing_ for you? It's quite verbose and I thought it would convey the idea of what it's doing... :disappointed: